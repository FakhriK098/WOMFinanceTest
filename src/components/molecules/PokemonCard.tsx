import React, { useMemo } from 'react'
import { Image, StyleSheet, View, useColorScheme } from 'react-native'
import Text from '../atoms/Text'
import type { PokemonListItem } from '../../types/pokemon'
import { colors } from '../../theme/colors'
import { spacing } from '../../theme/spacing'
import { radius } from '../../theme/radius'
import { typography } from '../../theme/typography'
import Spacer from '../atoms/Spacer'
import { tintFromHex } from '../../utils/pokemon'

type Props = { item: PokemonListItem }

const PokemonCard = ({ item }: Props) => {
  const isDark = useColorScheme() === 'dark'
  const containerStyle = useMemo(
    () => [
      styles.card,
      {
        backgroundColor: tintFromHex(item.color, 0.15),
        borderColor: item.color,
      },
    ],
    [item.color]
  )
  const nameColor = isDark ? colors.text.dark : colors.text.light
  const typeColor = isDark ? colors.text.mutedDark : colors.text.mutedLight
  return (
    <View style={containerStyle}>
      {!!item.image && <Image source={{ uri: item.image }} style={styles.image} resizeMode="contain" />}
      <Spacer size={spacing.sm} />
      <Text weight="bold" size={typography.sm} color={nameColor}>
        {item.name}
      </Text>
      <Text size={typography.xs} color={typeColor}>
        {item.types.join(' â€¢ ')}
      </Text>
    </View>
  )
}

const styles = StyleSheet.create({
  card: {
    flex: 1,
    margin: spacing.sm / 2,
    padding: spacing.md,
    borderRadius: radius.lg,
    borderWidth: 1,
  },
  image: {
    width: '100%',
    height: 100,
  },
})

export default React.memo(PokemonCard)
