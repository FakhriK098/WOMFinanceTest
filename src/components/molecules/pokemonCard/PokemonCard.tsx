import React, { useMemo } from 'react';
import { Image, Pressable, StyleSheet, useColorScheme } from 'react-native';
import Text from '@components/atoms/text/Text';
import type { PokemonListItem } from '@typings/pokemon';
import { colors } from '@theme/colors';
import { spacing } from '@theme/spacing';
import { radius } from '@theme/radius';
import { typography } from '@theme/typography';
import Spacer from '@components/atoms/spacer/Spacer';
import { tintFromHex } from '@utils/pokemon';

type Props = { item: PokemonListItem; onPress?: () => void };

const PokemonCard = ({ item, onPress }: Props) => {
  const isDark = useColorScheme() === 'dark';
  const containerStyle = useMemo(
    () => [
      styles.card,
      {
        backgroundColor: tintFromHex(item.color, 0.15),
        borderColor: item.color,
      },
    ],
    [item.color],
  );
  const nameColor = isDark ? colors.text.dark : colors.text.light;
  const typeColor = isDark ? colors.text.mutedDark : colors.text.mutedLight;
  return (
    <Pressable style={containerStyle} onPress={onPress}>
      {!!item.image && (
        <Image
          source={{ uri: item.image }}
          style={styles.image}
          resizeMode="contain"
        />
      )}
      <Spacer size={spacing.sm} />
      <Text weight="bold" size={typography.sm} color={nameColor}>
        {item.name}
      </Text>
      <Text size={typography.xs} color={typeColor}>
        {item.types.join(' â€¢ ')}
      </Text>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  card: {
    flex: 1,
    margin: spacing.sm / 2,
    padding: spacing.md,
    borderRadius: radius.lg,
    borderWidth: 1,
  },
  image: {
    width: '100%',
    height: 100,
  },
});

export default React.memo(PokemonCard);
