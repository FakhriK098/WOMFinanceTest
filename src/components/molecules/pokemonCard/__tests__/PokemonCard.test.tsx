import React from 'react';
import renderer, { act } from 'react-test-renderer';

import PokemonCard from '../PokemonCard';
import * as RN from 'react-native';

const item = {
  id: 1,
  name: 'Bulbasaur',
  image: 'http://img',
  types: ['grass', 'poison'],
  color: '#22c55e',
};

describe('PokemonCard molecule', () => {

  it('matches snapshot', () => {
    jest.spyOn(RN, 'useColorScheme').mockReturnValue('light');
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(<PokemonCard item={item as any} />);
    });
    expect(tr!.toJSON()).toMatchSnapshot();
    act(() => tr!.unmount());
  });

  it('renders name, types and calls onPress', () => {
    jest.spyOn(RN, 'useColorScheme').mockReturnValue('light');
    const onPress = jest.fn();
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(<PokemonCard item={item as any} onPress={onPress} />);
    });
    const nameNode = tr!.root.find(
      (inst) => inst.props && inst.props.children === 'Bulbasaur',
    );
    expect(nameNode).toBeTruthy();
    const typesNode = tr!.root.find(
      (inst) => inst.props && inst.props.children === 'grass â€¢ poison',
    );
    expect(typesNode).toBeTruthy();
    const pressable = tr!.root.find((inst) => inst.props && typeof inst.props.onPress === 'function');
    act(() => {
      pressable.props.onPress();
    });
    expect(onPress).toHaveBeenCalled();
    act(() => tr!.unmount());
  });
});
