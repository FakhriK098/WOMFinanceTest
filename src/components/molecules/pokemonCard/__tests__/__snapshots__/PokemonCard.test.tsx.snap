// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard molecule matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "borderRadius": 16,
        "borderWidth": 1,
        "flex": 1,
        "margin": 4,
        "padding": 14,
      },
      {
        "backgroundColor": "rgba(34, 197, 94, 0.15)",
        "borderColor": "#22c55e",
      },
    ]
  }
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "http://img",
      }
    }
    style={
      {
        "height": 100,
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "height": 8,
        "width": 8,
      }
    }
  />
  <Text
    style={
      [
        {
          "color": "#111827",
          "fontSize": 14,
          "fontWeight": "700",
        },
        undefined,
      ]
    }
  >
    Bulbasaur
  </Text>
  <Text
    style={
      [
        {
          "color": "#6B7280",
          "fontSize": 12,
          "fontWeight": "400",
        },
        undefined,
      ]
    }
  >
    grass â€¢ poison
  </Text>
</View>
`;
